File: app/src/main/java/com/theone/simpleshare/ui/pairing/PairingFragment.java
package com.theone.simpleshare.ui.pairing;



import static com.theone.simpleshare.bluetooth.BluetoothUtils.isA2dpDevice;
import static com.theone.simpleshare.bluetooth.BluetoothUtils.isInputDevice;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothClass;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.theone.simpleshare.R;
import com.theone.simpleshare.bluetooth.BluetoothPairingService;
import com.theone.simpleshare.databinding.FragmentPairingBinding;

import java.util.ArrayList;

public class PairingFragment extends Fragment {

    private PairingViewModel pairingViewModel;
    private FragmentPairingBinding binding;
    private final static String TAG = "PairingFragment";

    private Context mContext;
    private RecyclerView mRecyclerView;

    private RecyclerAdapter mRecyclerAdapter;

    private enum ParingMode {AUTO_PAIRING_MODE, MANUAL_PAIRING_MODE}

    private ParingMode mPairingMode = ParingMode.MANUAL_PAIRING_MODE;



    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        pairingViewModel =
                new ViewModelProvider(this).get(PairingViewModel.class);

        binding = FragmentPairingBinding.inflate(inflater, container, false);
        View root = binding.getRoot();
        mContext = getActivity();



        pairingViewModel.getList().observe(getViewLifecycleOwner(), new Observer<ArrayList<Item>>() {
            @Override
            public void onChanged(ArrayList<Item> items) {
                Log.d(TAG, "onChanged");
            }
        });
        mRecyclerAdapter = new RecyclerAdapter();
        mRecyclerView = binding.recyclerView;
        mRecyclerView.setLayoutManager(new LinearLayoutManager(mContext));
        mRecyclerView.setLayoutManager(new LinearLayoutManager(mContext, RecyclerView.VERTICAL, false));
        mRecyclerView.setAdapter(mRecyclerAdapter);

        mRecyclerAdapter.setOnItemClickListener(new RecyclerAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View v, int pos, Item item) {
                //Toast.makeText(mContext, " pos : "+pos , Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(mContext, BluetoothPairingService.class);

                intent.setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_PAIRING);
                intent.putExtra(BluetoothDevice.EXTRA_DEVICE, item.device);

                Toast.makeText(mContext, "start pairing service",Toast.LENGTH_LONG).show();
                mContext.startService(intent);

            }
        });

        mRecyclerAdapter.setItemList(pairingViewModel.getList().getValue());
        binding.manualPairing.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, BluetoothPairingService.class);
                intent.setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_SCAN);
                Toast.makeText(mContext, "start scan",Toast.LENGTH_LONG).show();

                mPairingMode = ParingMode.MANUAL_PAIRING_MODE;

                mContext.startService(intent);

                /*refresh rv*/
                mRecyclerAdapter.clear();
            }
        });
        binding.autoPairing.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, BluetoothPairingService.class);
                intent.setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_SCAN);


                Toast.makeText(mContext, "start auto pairing",Toast.LENGTH_LONG).show();

                mPairingMode = ParingMode.AUTO_PAIRING_MODE;
                mContext.startService(intent);
                /*refresh rv*/
                mRecyclerAdapter.clear();

            }
        });
        binding.bondedDevice.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                drawBondedDevices();
            }
        });
        return root;
    }

    private void drawBondedDevices(){
        BluetoothManager bluetoothManager = (BluetoothManager) getActivity().getSystemService(Context.BLUETOOTH_SERVICE);
        BluetoothAdapter bluetoothAdapter = bluetoothManager.getAdapter();

        /*refresh rv*/
        mRecyclerAdapter.clear();

        for(BluetoothDevice device :bluetoothAdapter.getBondedDevices()){
            pairingViewModel.getList().getValue().add(new Item(R.drawable.ic_launcher_foreground, device.getName(),
                    device.getAddress(), device));

        }
    }

    private final BroadcastReceiver mBroadcast = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {

            String action = intent.getAction();

            if (action != null) {
                Log.d(TAG, "Processing " + action);
            }
            switch (action) {

                case BluetoothDevice.ACTION_FOUND :
                    BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
                    if (device != null) {
                        /*Deal with HID and A2DP device*/
                        if ( isA2dpDevice(device)|| isInputDevice(device) ) {
                            Log.d(TAG, "device : " + device.getName());
                            pairingViewModel.getList().getValue().add( new Item(R.drawable.ic_launcher_foreground, device.getName(),
                                    device.getAddress(), device));
                            mRecyclerAdapter.notifyItemInserted(pairingViewModel.getList().getValue().size());

                            if(mPairingMode == ParingMode.AUTO_PAIRING_MODE){
                                Toast.makeText(mContext, "AUTO PAIR :" +device.getName(),Toast.LENGTH_LONG).show();
                                Log.d(TAG, "AUTO PAIR :" +device.getName());

                                Intent i = new Intent(mContext, BluetoothPairingService.class);
                                i.setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_PAIRING);
                                i.putExtra(BluetoothDevice.EXTRA_DEVICE, device);

                                mContext.startService(i);
                            }
                        } else {
                            if( device.getName() != null) {
                                pairingViewModel.getList().getValue().add(new Item(R.drawable.ic_launcher_foreground, device.getName(),
                                        device.getAddress(), device));
                                mRecyclerAdapter.notifyItemInserted(pairingViewModel.getList().getValue().size());
                            } else {
                                Log.w(TAG, "empty device name");
                            }
                        }

                    }

                    break;

                default:
                    Log.e(TAG, "onReceive: Error: unhandled action=" + action);
            }
        }
    };

    @Override
    public void onResume() {
        super.onResume();
        registerReceiver();
    }

    private void registerReceiver(){
        IntentFilter filter = new IntentFilter();
        filter.addAction(BluetoothDevice.ACTION_FOUND);
        mContext.registerReceiver(mBroadcast, filter);
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
        mContext.unregisterReceiver(mBroadcast);
    }
}

File: app/src/main/kotlin/com/theone/simpleshare/ui/pairing/PairingFragment.kt
package com.theone.simpleshare.ui.pairing

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.theone.simpleshare.MainActivity
import com.theone.simpleshare.R
import com.theone.simpleshare.bluetooth.BluetoothPairingService
import com.theone.simpleshare.bluetooth.BluetoothUtils.getResourceIcon
import com.theone.simpleshare.bluetooth.BluetoothUtils.isA2dpDevice
import com.theone.simpleshare.bluetooth.BluetoothUtils.isInputDevice
import com.theone.simpleshare.databinding.FragmentPairingBinding
import com.theone.simpleshare.viewmodel.Item
import com.theone.simpleshare.viewmodel.ItemViewModel
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.qualifiers.ActivityContext
import io.reactivex.internal.schedulers.IoScheduler
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import java.lang.Thread.sleep
import java.util.*

//https://developer88.tistory.com/349
@AndroidEntryPoint
class PairingFragment : Fragment() {
    private val itemViewModel: ItemViewModel by viewModels()
    private lateinit var binding: FragmentPairingBinding
    private lateinit var mContext: Context
    private lateinit var mRecyclerView: RecyclerView
    private lateinit var mRecyclerAdapter: RecyclerAdapter

    companion object {
        private const val TAG = "PairingFragment"
    }
    enum class ParingMode {
        AUTO_PAIRING_MODE, MANUAL_PAIRING_MODE
    }

    private var mPairingMode = ParingMode.MANUAL_PAIRING_MODE
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentPairingBinding.inflate(inflater, container, false)
        val root: View = binding.root
        mContext = activity as Context

        //test
        itemViewModel.deleteAll()

        mRecyclerAdapter = RecyclerAdapter()
        mRecyclerView = binding.recyclerView
        with(mRecyclerView) {
            setLayoutManager(LinearLayoutManager(mContext))
            setLayoutManager(LinearLayoutManager(mContext, RecyclerView.VERTICAL, false))
            setAdapter(mRecyclerAdapter)
        }
        mRecyclerAdapter.setOnItemClickListener (
            object : RecyclerAdapter.OnItemClickListener {
                override fun onItemClick(v: View?, pos: Int, item: Item) {
                    //Toast.makeText(mContext, " pos : "+pos , Toast.LENGTH_SHORT).show();
                    val intent = Intent(mContext, BluetoothPairingService::class.java)
                    intent.apply {
                        setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_PAIRING)
                        putExtra(BluetoothDevice.EXTRA_DEVICE, item.device)
                        Toast.makeText(mContext, "start pairing service", Toast.LENGTH_LONG).show()
                        mContext.startService(this)
                    }

                }
            })
        val data =  itemViewModel.getItemList().observe(this){ list ->

            mRecyclerAdapter.setItemList( list as ArrayList<Item>?)

        }

//        if ( itemViewModel.getItemList().value ==null){
//            Log.d("easy", "item list is empty" )
//            var list = ArrayList<Item>()
//            val device:BluetoothDevice? = null
//            list.add(Item(1,0,"empty","empty",device, -1,-1))
//            mRecyclerAdapter.setItemList(list)
//        } else {
//            mRecyclerAdapter.setItemList(itemViewModel.getItemList().value as ArrayList<Item>?)
//        }
        with(binding) {
            manualPairing.setOnClickListener {
                val intent = Intent(mContext, BluetoothPairingService::class.java)
                with(intent) {
                    setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_SCAN)
                    Toast.makeText(mContext, "start scan", Toast.LENGTH_LONG).show()
                    mPairingMode = ParingMode.MANUAL_PAIRING_MODE
                    mContext.startService(this)
                }
                /*refresh rv*/
                mRecyclerAdapter.clear()
            }
            autoPairing.setOnClickListener {
                val intent = Intent(mContext, BluetoothPairingService::class.java)
                with(intent) {
                    setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_SCAN)
                    mContext.startService(this)
                }

                Toast.makeText(mContext, "start auto pairing", Toast.LENGTH_LONG).show()
                mPairingMode = ParingMode.AUTO_PAIRING_MODE
                /*refresh rv*/
                mRecyclerAdapter.clear()
            }
            bondedDevice.setOnClickListener {
                drawBondedDevices()
            }

            roomDbTest.setOnClickListener {
              val fruits = listOf("apple","banna","kiwi","cherry")
              val res=  fruits.asSequence().filter {

                    Log.d(TAG, "checking length of $it")
                    it.length > 2
                }
                    .map {
                        Log.d(TAG, "mapping to the length of $it")
                        "${it.length}"
                    }.take(2)


                Log.d(TAG,"" +res.toList() )



            }
        }
        return root
    }

    private fun drawBondedDevices() {
        val bluetoothManager: BluetoothManager =
            mContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        val bluetoothAdapter: BluetoothAdapter = bluetoothManager.getAdapter()

        with(bluetoothAdapter){
            if(isDiscovering)
                cancelDiscovery();
        }

        /*refresh rv*/
        mRecyclerAdapter.clear()
        for (device in bluetoothAdapter.getBondedDevices()) {
            itemViewModel.insertItem(
                Item(1,
                    R.drawable.ic_launcher_foreground, device.getName(),
                    device.getAddress(), device, -1,-1)
            )
        }
    }
    var itemId = 100
    private val mBroadcast: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action: String? = intent.action
            if (action != null) {
                Log.d(TAG, "Processing $action")
            }

            when (action) {
                BluetoothDevice.ACTION_FOUND -> {
                    val device: BluetoothDevice? =
                        intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)

                    device?.let { device ->
                        if (isA2dpDevice(device) || isInputDevice(device)) {
                            Log.d(TAG, "device : " + device.getName())
                            itemViewModel.insertItem(
                                Item(itemId++,
                                    getResourceIcon(device), device.getName(),
                                    device.getAddress(), device, -1,-1
                                )
                            )
                             itemViewModel.getItemList().observe(this@PairingFragment){
                                with(mRecyclerAdapter){
                                    setItemList( it as ArrayList<Item>?)
                                    notifyItemInserted( it.size - 1 )
                                }
                            }

                            if (mPairingMode == ParingMode.AUTO_PAIRING_MODE) {
                                Toast.makeText(
                                    mContext,
                                    "AUTO PAIR :" + device.getName(),
                                    Toast.LENGTH_LONG
                                ).show()
                                Log.d(TAG, "AUTO PAIR :" + device.getName())
                                val i = Intent(mContext, BluetoothPairingService::class.java)
                                i.setAction(BluetoothPairingService.BLUETOOTH_ACTION_START_PAIRING)
                                i.putExtra(BluetoothDevice.EXTRA_DEVICE, device)
                                mContext.startService(i)
                            }

                        }
                    }
                }
                else -> Log.e(TAG, "onReceive: Error: unhandled action=$action")
            }
        }
    }

    override fun onResume() {
        super.onResume()
        registerReceiver()
    }

    private fun registerReceiver() {
        val filter = IntentFilter()
        filter.addAction(BluetoothDevice.ACTION_FOUND)
        mContext.registerReceiver(mBroadcast, filter)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mContext.unregisterReceiver(mBroadcast)
    }

}